File: frontend/src/lib/chem-logic/strategist.ts
export interface StrategyRecommendation {
    type: string;
    recommended_ph: number;
    recommended_buffer_id: string;
    recommended_concentration: number;
    recommended_salt: number;
    salt_reason: string;
    recommended_additives: string[];
    additives_reason: string;
    reason: string;
    buffer_reason: string;
    concentration_reason: string;
    resin_type: string;
}

export interface PurificationStrategy {
    pI: number;
    unstable_zone: [number, number];
    strategies: StrategyRecommendation[];
}

export class PurificationStrategist {
    pi: number;
    sequence: string;

    constructor(pi: number, sequence: string = "") {
        this.pi = pi;
        this.sequence = sequence;
    }

    suggestStrategy(): PurificationStrategy {
        // Define Unstable Zone (Precipitation Risk)
        const unstable_zone: [number, number] = [
            parseFloat((this.pi - 0.5).toFixed(2)),
            parseFloat((this.pi + 0.5).toFixed(2))
        ];

        const strategy: PurificationStrategy = {
            pI: parseFloat(this.pi.toFixed(2)),
            unstable_zone: unstable_zone,
            strategies: []
        };

        // Cystiene check for additives
        const additives: string[] = [];
        let additives_reason = "";
        if (this.sequence && this.sequence.includes("C")) {
            additives.push("DTT");
            additives_reason = "Cysteines detected. Reducing agent (DTT) recommended to prevent aggregation.";
        }

        // Logic
        if (this.pi < 6.0) {
            // Acidic: Use Anion Exchange
            const target_ph = this.pi + 1.0;
            const rec_buffer = target_ph > 7.5 ? "tris" : (target_ph > 5.8 ? "phosphate" : "mes");

            strategy.strategies.push({
                type: "Anion Exchange",
                recommended_ph: parseFloat(target_ph.toFixed(2)),
                recommended_buffer_id: rec_buffer,
                recommended_concentration: 20,
                recommended_salt: 0,
                salt_reason: "Start at 0 mM NaCl to promote binding to IEX resin.",
                recommended_additives: additives,
                additives_reason: additives_reason,
                reason: "Protein is acidic (pI < 6.0). Needs pH > pI to be negatively charged for AEX.",
                buffer_reason: `Chosen for optimal buffering capacity at pH ${target_ph.toFixed(2)}.`,
                concentration_reason: "Standard starting ionic strength for IEX to allow binding.",
                resin_type: "Positive Ligand (e.g., Q-Sepharose)"
            });
        } else if (this.pi > 8.0) {
            // Basic: Use Cation Exchange
            const target_ph = this.pi - 1.0;
            const rec_buffer = target_ph > 5.8 ? "phosphate" : (target_ph > 5.5 ? "mes" : "acetate");

            strategy.strategies.push({
                type: "Cation Exchange",
                recommended_ph: parseFloat(target_ph.toFixed(2)),
                recommended_buffer_id: rec_buffer,
                recommended_concentration: 20,
                recommended_salt: 0,
                salt_reason: "Start at 0 mM NaCl to promote binding to IEX resin.",
                recommended_additives: additives,
                additives_reason: additives_reason,
                reason: "Protein is basic (pI > 8.0). Needs pH < pI to be positively charged for CEX.",
                buffer_reason: `Chosen for optimal buffering capacity at pH ${target_ph.toFixed(2)}.`,
                concentration_reason: "Standard starting ionic strength for IEX to allow binding.",
                resin_type: "Negative Ligand (e.g., SP-Sepharose)"
            });
        } else {
            // Neutral: Propose both
            const aex_ph = this.pi + 1.0;
            const cex_ph = this.pi - 1.0;

            // Option 1: AEX
            if (aex_ph <= 14.0) {
                const rec_buffer_aex = aex_ph > 7.5 ? "tris" : "phosphate";
                strategy.strategies.push({
                    type: "Anion Exchange",
                    recommended_ph: parseFloat(aex_ph.toFixed(2)),
                    recommended_buffer_id: rec_buffer_aex,
                    recommended_concentration: 20,
                    recommended_salt: 0,
                    salt_reason: "Start at 0 mM NaCl to promote binding to IEX resin.",
                    recommended_additives: additives,
                    additives_reason: additives_reason,
                    reason: "Neutral pI. AEX option (pH > pI).",
                    buffer_reason: `Chosen for optimal buffering capacity at pH ${aex_ph.toFixed(2)}.`,
                    concentration_reason: "Standard starting ionic strength for IEX to allow binding.",
                    resin_type: "Positive Ligand"
                });
            }

            // Option 2: CEX
            if (cex_ph >= 0.0) {
                const rec_buffer_cex = cex_ph > 5.8 ? "phosphate" : "mes";
                strategy.strategies.push({
                    type: "Cation Exchange",
                    recommended_ph: parseFloat(cex_ph.toFixed(2)),
                    recommended_buffer_id: rec_buffer_cex,
                    recommended_concentration: 20,
                    recommended_salt: 0,
                    salt_reason: "Start at 0 mM NaCl to promote binding to IEX resin.",
                    recommended_additives: additives,
                    additives_reason: additives_reason,
                    reason: "Neutral pI. CEX option (pH < pI).",
                    buffer_reason: `Chosen for optimal buffering capacity at pH ${cex_ph.toFixed(2)}.`,
                    concentration_reason: "Standard starting ionic strength for IEX to allow binding.",
                    resin_type: "Negative Ligand"
                });
            }
        }

        return strategy;
    }
}

File: backend/modules/strategist.py
class PurificationStrategist:
    def __init__(self, pi: float, sequence: str = None):
        self.pi = pi
        self.sequence = sequence

    def suggest_strategy(self):
        """
        Suggests purification strategy based on pI.
        
        Logic:
        - pH_start = pI
        - Unstable Zone: pI +/- 0.5
        - Acidic Protein (pI < 6.0): Anion Exchange (AEX) at pH > pI (Net Charge Negative)
        - Basic Protein (pI > 8.0): Cation Exchange (CEX) at pH < pI (Net Charge Positive)
        - Neutral (6.0 - 8.0): Options.
        """
        
        # Define Unstable Zone (Precipitation Risk)
        unstable_zone = (round(self.pi - 0.5, 2), round(self.pi + 0.5, 2))
        
        strategy = {
            "pI": round(self.pi, 2),
            "unstable_zone": unstable_zone,
            "strategies": []
        }

        # Strategy Logic
        if self.pi < 6.0:
            # Acidic: Use Anion Exchange
            target_ph = self.pi + 1.0
            
            # Recommend Buffer
            rec_buffer = "tris" if target_ph > 7.5 else "phosphate" if target_ph > 5.8 else "mes"

            # Recommend Additives based on sequence
            additives = []
            additives_reason = ""
            if self.sequence and "C" in self.sequence:
                additives.append("DTT")
                additives_reason = "Cysteines detected. Reducing agent (DTT) recommended to prevent aggregation."

            strategy["strategies"].append({
                "type": "Anion Exchange",
                "recommended_ph": round(target_ph, 2),
                "recommended_buffer_id": rec_buffer,
                "recommended_concentration": 20,
                "recommended_salt": 0,
                "salt_reason": "Start at 0 mM NaCl to promote binding to IEX resin.",
                "recommended_additives": additives,
                "additives_reason": additives_reason,
                "reason": "Protein is acidic (pI < 6.0). Needs pH > pI to be negatively charged for AEX.",
                "buffer_reason": f"Chosen for optimal buffering capacity at pH {round(target_ph, 2)}.",
                "concentration_reason": "Standard starting ionic strength for IEX to allow binding.",
                "resin_type": "Positive Ligand (e.g., Q-Sepharose)"
            })
        elif self.pi > 8.0:
            # Basic: Use Cation Exchange
            target_ph = self.pi - 1.0
            
            # Recommend Buffer
            rec_buffer = "phosphate" if target_ph > 5.8 else "mes" if target_ph > 5.5 else "acetate" 

            # Recommend Additives based on sequence
            additives = []
            additives_reason = ""
            if self.sequence and "C" in self.sequence:
                additives.append("DTT")
                additives_reason = "Cysteines detected. Reducing agent (DTT) recommended to prevent aggregation."
            
            strategy["strategies"].append({
                "type": "Cation Exchange",
                "recommended_ph": round(target_ph, 2),
                "recommended_buffer_id": rec_buffer,
                "recommended_concentration": 20,
                 "recommended_salt": 0,
                "salt_reason": "Start at 0 mM NaCl to promote binding to IEX resin.",
                "recommended_additives": additives,
                "additives_reason": additives_reason,
                "reason": "Protein is basic (pI > 8.0). Needs pH < pI to be positively charged for CEX.",
                "buffer_reason": f"Chosen for optimal buffering capacity at pH {round(target_ph, 2)}.",
                "concentration_reason": "Standard starting ionic strength for IEX to allow binding.",
                "resin_type": "Negative Ligand (e.g., SP-Sepharose)"
            })
        else:
            # Neutral: Propose both with caution
            aex_ph = self.pi + 1.0
            cex_ph = self.pi - 1.0
            
             # Recommend Additives based on sequence
            additives = []
            additives_reason = ""
            if self.sequence and "C" in self.sequence:
                additives.append("DTT")
                additives_reason = "Cysteines detected. Reducing agent (DTT) recommended to prevent aggregation."

            # Option 1: AEX
            if aex_ph <= 14.0:
                rec_buffer_aex = "tris" if aex_ph > 7.5 else "phosphate"
                strategy["strategies"].append({
                    "type": "Anion Exchange",
                    "recommended_ph": round(aex_ph, 2),
                    "recommended_buffer_id": rec_buffer_aex,
                    "recommended_concentration": 20,
                    "recommended_salt": 0,
                    "salt_reason": "Start at 0 mM NaCl to promote binding to IEX resin.",
                    "recommended_additives": additives,
                    "additives_reason": additives_reason,
                    "reason": "Neutral pI. AEX option (pH > pI).",
                    "buffer_reason": f"Chosen for optimal buffering capacity at pH {round(aex_ph, 2)}.",
                    "concentration_reason": "Standard starting ionic strength for IEX to allow binding.",
                    "resin_type": "Positive Ligand"
                })
            
            # Option 2: CEX
            if cex_ph >= 0.0:
                rec_buffer_cex = "phosphate" if cex_ph > 5.8 else "mes"
                strategy["strategies"].append({
                    "type": "Cation Exchange",
                    "recommended_ph": round(cex_ph, 2),
                     "recommended_buffer_id": rec_buffer_cex,
                    "recommended_concentration": 20,
                    "recommended_salt": 0,
                    "salt_reason": "Start at 0 mM NaCl to promote binding to IEX resin.",
                    "recommended_additives": additives,
                    "additives_reason": additives_reason,
                    "reason": "Neutral pI. CEX option (pH < pI).",
                    "buffer_reason": f"Chosen for optimal buffering capacity at pH {round(cex_ph, 2)}.",
                    "concentration_reason": "Standard starting ionic strength for IEX to allow binding.",
                    "resin_type": "Negative Ligand"
                })
                
        return strategy
